In [1]:
%cd ..

D:\SoftUni\Data Science\ProjectV2


### Imports

In [2]:
import pandas as pd

import numpy as np

from src import functions

from functools import reduce

### Metadata

The data about the rates of different mental illnesses includes information about Schizophrenia, Anxiety, Attention Deficit / Hyperactivity, Bipolar and Depressive disorders. It is age - standardized which means that it takes into account the differences in the age structure of each country. [[1]](#references) The rates are provided in the 'Value' column and are measured in prevalent cases per 100 000 people. Thus the values ​​of different countries could be compared unbiased.

The data is saved in five separate datasets, each with 204 observations of 9 features.

The source of data is https://www.healthdata.org/research-analysis/health-risks-issues/mental-health.

### Loading data

In [28]:
amphetamine_use_disorders_data = pd.read_csv('data/suds_source_datasets/amphetamine_use_disorders.csv')

In [29]:
amphetamine_use_disorders_data.head(2)

Location    Year               Age   Sex   Cause of death or injury  \
0  Afghanistan  2021.0  Age-standardized  Both  Amphetamine use disorders   
1       Angola  2021.0  Age-standardized  Both  Amphetamine use disorders   

                       Measure      Value  Lower bound  Upper bound  
0  Prevalent cases per 100,000  31.959143    21.495979    44.618085  
1  Prevalent cases per 100,000  38.395224    26.234020    53.864415

In [5]:
amphetamine_use_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [6]:
cannabis_use_disorders_data = pd.read_csv('data/suds_source_datasets/cannabis_use_disorders.csv')
cannabis_use_disorders_data.head(2)

Location    Year               Age   Sex Cause of death or injury  \
0  Afghanistan  2021.0  Age-standardized  Both   Cannabis use disorders   
1       Angola  2021.0  Age-standardized  Both   Cannabis use disorders   

                       Measure       Value  Lower bound  Upper bound  
0  Prevalent cases per 100,000  129.592702    87.148175   189.891116  
1  Prevalent cases per 100,000  177.467520   121.765687   262.167628

In [7]:
cannabis_use_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [8]:
cocaine_use_disorders_data = pd.read_csv('data/suds_source_datasets/cocaine_use_disorders.csv')

In [9]:
cocaine_use_disorders_data.head(2)

Location    Year               Age   Sex Cause of death or injury  \
0  Afghanistan  2021.0  Age-standardized  Both    Cocaine use disorders   
1       Angola  2021.0  Age-standardized  Both    Cocaine use disorders   

                       Measure      Value  Lower bound  Upper bound  
0  Prevalent cases per 100,000  13.949386     9.618930    19.190581  
1  Prevalent cases per 100,000   8.157492     6.023403    10.683326

In [10]:
cocaine_use_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


In [11]:
opioid_use_disorders_data = pd.read_csv('data/suds_source_datasets/opioid_use_disorders.csv')

In [12]:
opioid_use_disorders_data.head(2)

Location    Year               Age   Sex Cause of death or injury  \
0  Afghanistan  2021.0  Age-standardized  Both     Opioid use disorders   
1       Angola  2021.0  Age-standardized  Both     Opioid use disorders   

                       Measure       Value  Lower bound  Upper bound  
0  Prevalent cases per 100,000  248.667360   213.530375   285.921175  
1  Prevalent cases per 100,000   72.370541    59.067434    87.497248

In [13]:
opioid_use_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Location                  206 non-null    object 
 1   Year                      204 non-null    float64
 2   Age                       204 non-null    object 
 3   Sex                       204 non-null    object 
 4   Cause of death or injury  204 non-null    object 
 5   Measure                   204 non-null    object 
 6   Value                     204 non-null    float64
 7   Lower bound               204 non-null    float64
 8   Upper bound               204 non-null    float64
dtypes: float64(4), object(5)
memory usage: 14.6+ KB


All datasets have the same shape - 206 rows and 9 columns. They don't contain missing values.

### Feature selection

In [14]:
opioid_use_disorders_data.nunique()

Location                    206
Year                          1
Age                           1
Sex                           1
Cause of death or injury      1
Measure                       1
Value                       204
Lower bound                 204
Upper bound                 204
dtype: int64

All observations in the above datasets are age standardized for 2021. The values of Substance uce disorders are measured by a common unit - Prevalent cases per 100 000. So, each of the columns 'Year', 'Age', 'Sex', 'Cause of death and injury' and 'Measure' has a single unique value for all observations. I will drop them as they don't add any useful value. 'Lower bound' and 'Upper bound' columns also will be dropped as they are not features of interest of this project.

In [15]:
columns_to_drop = ['Lower bound', 'Upper bound', 'Year', 'Age', 'Sex', 'Cause of death or injury', 'Measure']

In [16]:
amphetamine_use_disorders_data.drop(columns = columns_to_drop, inplace = True)
cannabis_use_disorders_data.drop(columns = columns_to_drop, inplace = True)
cocaine_use_disorders_data.drop(columns = columns_to_drop, inplace = True)
opioid_use_disorders_data.drop(columns = columns_to_drop, inplace = True)

### Merging the data about mental disorders rates into a common dataset

In [17]:
# The 'Value' columns will be renamed for better description of the features. 'Location' column will be renamed to 'country' for consistency with the other dataset.
amphetamine_use_disorders_data = amphetamine_use_disorders_data.rename(columns = {'Value': 'amphetamine_disorders_rate', 'Location': 'country'})
cannabis_use_disorders_data = cannabis_use_disorders_data.rename(columns = {'Value': 'cannabis_disorders_rate', 'Location': 'country'})
cocaine_use_disorders_data = cocaine_use_disorders_data.rename(columns = {'Value': 'cocaine_disorders_rate', 'Location': 'country'})
opioid_use_disorders_data = opioid_use_disorders_data.rename(columns = {'Value': 'opioid_disorders_rate', 'Location': 'country'})

In [18]:
suds_disorders_data = reduce(
    lambda x, y: pd.merge(x, y, on = 'country', how = 'inner'), [
        amphetamine_use_disorders_data, 
        cannabis_use_disorders_data, 
        cocaine_use_disorders_data, 
        opioid_use_disorders_data, 
    ]
)

### Exploration and cleaning of resulted dataset

In [19]:
suds_disorders_data

country  \
0                                          Afghanistan   
1                                               Angola   
2                                              Albania   
3                                              Andorra   
4                                 United Arab Emirates   
..                                                 ...   
201                                              Egypt   
202                                              Sudan   
203                                              China   
204  Institute for Health Metrics and Evaluation (I...   
205  For terms and conditions of use, please visit ...   

     amphetamine_disorders_rate  cannabis_disorders_rate  \
0                     31.959143               129.592702   
1                     38.395224               177.467520   
2                    130.598115               246.323973   
3                    130.148605               468.433635   
4                     53.148291               138.318626   
..                          ...                      ...   
201                   34.207386               112.056301   
202                   32.240389               129.055571   
203                  267.200599               206.552144   
204                         NaN                      NaN   
205                         NaN                      NaN   

     cocaine_disorders_rate  opioid_disorders_rate  
0                 13.949386             248.667360  
1                  8.157492              72.370541  
2                 28.224969             133.757603  
3                129.317154             111.674840  
4                 17.739865             371.353942  
..                      ...                    ...  
201               11.704605             199.131103  
202               13.291759             177.965073  
203                6.152595              94.350923  
204                     NaN                    NaN  
205                     NaN                    NaN  

[206 rows x 5 columns]

In [20]:
suds_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206 entries, 0 to 205
Data columns (total 5 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   country                     206 non-null    object 
 1   amphetamine_disorders_rate  204 non-null    float64
 2   cannabis_disorders_rate     204 non-null    float64
 3   cocaine_disorders_rate      204 non-null    float64
 4   opioid_disorders_rate       204 non-null    float64
dtypes: float64(4), object(1)
memory usage: 8.2+ KB


In [21]:
suds_disorders_data.duplicated().unique()

array([False])

The resulted dataset has 206 rows and 6 features.

There are no duplicated observations nor missing values.

The last row is not observatios but contains information about data provider and terms of use. It is safe to delete it.

In [22]:
suds_disorders_data = suds_disorders_data.iloc[:-1]

In [23]:
suds_disorders_data.tail(2)

country  \
203                                              China   
204  Institute for Health Metrics and Evaluation (I...   

     amphetamine_disorders_rate  cannabis_disorders_rate  \
203                  267.200599               206.552144   
204                         NaN                      NaN   

     cocaine_disorders_rate  opioid_disorders_rate  
203                6.152595              94.350923  
204                     NaN                    NaN

In [24]:
# The float values will be rounded for better readability.
suds_disorders_data = suds_disorders_data.round(0)

In [25]:
suds_disorders_data.head(3)

country  amphetamine_disorders_rate  cannabis_disorders_rate  \
0  Afghanistan                        32.0                    130.0   
1       Angola                        38.0                    177.0   
2      Albania                       131.0                    246.0   

   cocaine_disorders_rate  opioid_disorders_rate  
0                    14.0                  249.0  
1                     8.0                   72.0  
2                    28.0                  134.0

In [26]:
suds_disorders_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 5 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   country                     205 non-null    object 
 1   amphetamine_disorders_rate  204 non-null    float64
 2   cannabis_disorders_rate     204 non-null    float64
 3   cocaine_disorders_rate      204 non-null    float64
 4   opioid_disorders_rate       204 non-null    float64
dtypes: float64(4), object(1)
memory usage: 8.1+ KB


The dtypes of features in resulted dataframe are appropriate so no need to transform them.

### Saving transformed data
The resulted dataset is saved for further use in the project.

In [27]:
suds_disorders_data.to_csv('data/preprocessed_datasets/sud_disorders_by_country.csv', index = False)

<a id='references'></a>

### References
1. [Age standardization - ourworldindata.org](https://ourworldindata.org/age-standardization)